File: 3647-zero-array-transformation-iii/zero-array-transformation-iii.cpp
class Solution {
public:
    int maxRemoval(vector<int>& nums, vector<vector<int>>& queries) {
        int n=nums.size();
        int q=queries.size();
        priority_queue<int> maxheap;
        priority_queue<int,vector<int>,greater<int>> past;
        sort(queries.begin(),queries.end());
        int j=0;
        int used=0;
        for(int i=0;i<n;i++){
            while(j<q && queries[j][0]==i){
                maxheap.push(queries[j][1]);
                j++;
            }
            nums[i]-=past.size();
            while(nums[i]>0 && !maxheap.empty() && maxheap.top()>=i){
                int ending=maxheap.top();
                maxheap.pop();
                past.push(ending);
                used++;
                nums[i]--;  
            }
            if(nums[i]>0){
                return -1;
            }
            while(!past.empty() && past.top()<=i){
                past.pop();
            }
            

        }
        return q-used;
    }
};

File: 3647-zero-array-transformation-iii/zero-array-transformation-iii.java
class Solution {
    public int maxRemoval(int[] nums, int[][] que) {
        int q=que.length;
        int n=nums.length;
        PriorityQueue<Integer> maxheap=new PriorityQueue<>((a,b)->b-a);
        PriorityQueue<Integer> past=new PriorityQueue<>();
        Arrays.sort(que,(a,b)->a[0]-b[0]);
        int j=0;
        int used=0;
        for(int i=0;i<n;i++){
            while(j<q && que[j][0]==i){
                maxheap.add(que[j][1]);
              //  System.out.println(que[j][1]);
                j++;
            }
            nums[i]-=past.size();
            while(!maxheap.isEmpty() && nums[i]>0 && maxheap.peek()>=i){
                int ending=maxheap.poll();
                nums[i]-=1;
                past.add(ending);
              //  System.out.println("he;;");
                used++;
            }
            if(nums[i]>0){
                return -1;
            }
            while(!past.isEmpty() && past.peek()<=i){
                past.poll();
            }
        }
        return q-used;
    }
}